src/LockTable.h

```diff
@@ -0,0 +1,187 @@
+ *
+ * The LockTable is an array of #buckets.  Locks are "hashed" into buckets based
+ * on the lower n-1 bits of the lockId (n is such that 2^n is the number of
+ * buckets); the implementation assumes that lockIds are relatively uniformly
+ * distributed.
+ *
+ * Each bucket consists of a one or more chained \link CacheLine
+ * CacheLines\endlink, the first of which lives inline in the array of buckets.
+ * Each cache line consists of several LockTable Entry objects; each Entry
+ * contains the lockId of an acquired lock or 0; only acquired locks are
+ * kept in the LockTable.  In exception, the first entry in a bucket's first
+ * cache line is specially designated and is cast and used as the bucket's
+ * \link BucketLock BucketLock\endlink; this is a monitor-style spin-lock that
+ * protects against concurrent accesses to the same bucket.
+ *
+ * If there are two many acquired locks to fit in the bucket's first cache line,
+ * additional overflow cache lines are allocated (outside the array of buckets)
+ * and chained together.  Cache lines are never removed from buckets; lock
+ * acquisition performance in a particular bucket degrades as more cache lines
+ * are added.
+ *
+ * For best performance, the number of buckets should be set large enough so
+ * that overflow cache lines are almost never needed but small enough that the
+ * entire structure might fit in CPU cache.
+ */
+class LockTable {
+  PUBLIC:
+    explicit LockTable(uint64_t numBuckets);
+    virtual ~LockTable();
+
+    void acquireLock(uint64_t lockId);
+    bool isLockAcquired(uint64_t lockId);
+    void releaseLock(uint64_t lockId);
+    bool tryAcquireLock(uint64_t lockId);
+
+  PRIVATE:
+    /**
+     * A lock table entry.
+     *
+     * Lock table entries live in \link CacheLine CacheLines\endlink.  A lock
+     * table entry holds the lockId of an acquired lock.
+     */
+    typedef uint64_t Entry;
+    static_assert(sizeof(Entry) == 8, "LockTable::Entry is not 8 bytes");
+
+    /**
+     * The number of bytes per cache line in this machine.
+     */
+    static const uint32_t BYTES_PER_CACHE_LINE = 64;
+
+    /**
+     * The number of LockTable Entry objects in a CacheLine.
+     */
+    static const uint32_t ENTRIES_PER_CACHE_LINE = (BYTES_PER_CACHE_LINE /
+                                                    sizeof(Entry)) - 1;
+    static_assert(BYTES_PER_CACHE_LINE % sizeof(Entry) == 0,
+                  "BYTES_PER_CACHE_LINE not a multiple of sizeof(Entry)");
+
+    /**
+     * A linked list of cache lines composes a bucket within the LockTable.
+     *
+     * Each cache line is composed of several lock table entries and a pointer
+     * to the next overflow cache line (if any).
+     *
+     * A CacheLine is meant to fit on a single L2 cache line on the CPU.
+     * Different processors may require tweaking ENTRIES_PER_CACHE_LINE to
+     * achieve this.
+     */
+    struct CacheLine {
+        /// Array of lock table entries (acquired lockIds).
```

Comment:
Would it make sense to break out the first entry and  declared specially, since it is actually a lock, not an entry? If nothing else, you should document here that the first entry is special.